extensions/arrow/src/main/java/io/deephaven/extensions/arrow/ArrowWrapperTools.java

```diff
@@ -70,162 +81,169 @@
+ * <li>{@link ArrowObjectColumnSource ArrowObjectColumnSource&lt;BigDecimal&gt;} - uses {@link Decimal256Vector} under the hood, returns BigDecimal</li>
+ * <li>{@link ArrowObjectColumnSource ArrowObjectColumnSource&lt;LocalDateTime&gt;} - uses {@link DateMilliVector} under the hood, returns LocalDateTime</li>
+ * </ul>
  * </pre>
  */
 public class ArrowWrapperTools {
+    private static final int MAX_POOL_SIZE = Math.max(UpdateGraphProcessor.DEFAULT.getUpdateThreads(),
+            Configuration.getInstance().getIntegerWithDefault("ArrowWrapperTools.defaultMaxPooledContext", 4));
+
     private static final BufferAllocator rootAllocator = new RootAllocator();
 
     /**
-     * Reads arrow data from file and returns a query table
+     * Reads arrow data from a feather-formatted file and returns a query table
      */
-    public static QueryTable readArrow(final @NotNull String path) {
-        final Helper helper = new Helper(path, rootAllocator);
-        final Shareable context = helper.newShareable();
-        final ArrowFileReader reader = context.getReader();
-        try {
+    public static QueryTable readFeather(final @NotNull String path) {
+        final ArrowTableContext arrowTableContext = new ArrowTableContext(path, rootAllocator);
+        try (final Shareable context = arrowTableContext.newShareable();
+                final SeekableByteChannel channel = arrowTableContext.openChannel()) {
+            final ArrowFileReader reader = context.getReader();
             int biggestBlock = 0;
             final List<ArrowBlock> recordBlocks = reader.getRecordBlocks();
             final int[] blocks = new int[recordBlocks.size()];
-            for (int i = 0; i < recordBlocks.size(); i++) {
-                reader.loadRecordBatch(recordBlocks.get(i));
-                final int schemaRowCount = reader.getVectorSchemaRoot().getRowCount();
-                blocks[i] = schemaRowCount;
-                if (schemaRowCount > biggestBlock) {
-                    biggestBlock = schemaRowCount;
+            // TODO: load only the metadata to speed up initial read time
+
+            int metadataBufLen = 1024;
+            byte[] rawMetadataBuf = new byte[metadataBufLen];
+            for (int ii = 0; ii < blocks.length; ++ii) {
+                final ArrowBlock block = recordBlocks.get(ii);
+                while (block.getMetadataLength() > metadataBufLen) {
+                    metadataBufLen *= 2;
+                    rawMetadataBuf = new byte[metadataBufLen];
                 }
+
+                channel.position(block.getOffset());
+
+                final ByteBuffer metadataBuf = ByteBuffer.wrap(rawMetadataBuf, 0, block.getMetadataLength());
+                int numRead = channel.read(metadataBuf);
+                if (numRead != block.getMetadataLength()) {
+                    throw new IOException("Unexpected end of input trying to read batch.");
```

Comment:
Maybe we could offer more information (file name, number read and expected, block index, mother's maiden name, etc).